xarray/core/missing.py

```diff
@@ -0,0 +1,330 @@
+                'recieved invalid kwargs: %r' % self.cons_kwargs.keys())
+
+        if fill_value is None:
+            self._left = np.nan
+            self._right = yi[-1]
+        elif isinstance(fill_value, Iterable) and len(fill_value) == 2:
+            self._left = fill_value[0]
+            self._right = fill_value[1]
+        elif is_scalar(fill_value):
+            self._left = fill_value
+            self._right = fill_value
+        else:
+            raise ValueError('%s is not a valid fill_value' % fill_value)
+
+    def __call__(self, x):
+        return self.f(x, self._xi, self._yi, left=self._left,
+                      right=self._right, **self.call_kwargs)
+
+
+class ScipyInterpolator(BaseInterpolator):
+    '''Interpolate a 1-D function using Scipy interp1d
+
+    See Also
+    --------
+    scipy.interpolate.interp1d
+    '''
+    def __init__(self, xi, yi, method=None, fill_value=None,
+                 assume_sorted=True, copy=False, bounds_error=False, **kwargs):
+        from scipy.interpolate import interp1d
+
+        if method is None:
+            raise ValueError('method is a required argument, please supply a '
+                             'valid scipy.inter1d method (kind)')
+
+        if method == 'polynomial':
+            method = kwargs.pop('order', None)
+            if method is None:
+                raise ValueError('order is required when method=polynomial')
+
+        self.method = method
+
+        self.cons_kwargs = kwargs
+        self.call_kwargs = {}
+
+        if fill_value is None and method == 'linear':
+            fill_value = kwargs.pop('fill_value', (np.nan, yi[-1]))
+        elif fill_value is None:
+            fill_value = np.nan
+
+        self.f = interp1d(xi, yi, kind=self.method, fill_value=fill_value,
+                          bounds_error=False, assume_sorted=assume_sorted,
+                          copy=copy, **self.cons_kwargs)
+
+
+class SplineInterpolator(BaseInterpolator):
+    '''One-dimensional smoothing spline fit to a given set of data points.
+
+    See Also
+    --------
+    scipy.interpolate.UnivariateSpline
+    '''
+    def __init__(self, xi, yi, method='spline', fill_value=None, order=3,
+                 **kwargs):
+        from scipy.interpolate import UnivariateSpline
+
+        if method != 'spline':
+            raise ValueError(
+                'only method `spline` is valid for the SplineInterpolator')
+
+        self.method = method
+        self.cons_kwargs = kwargs
+        self.call_kwargs['nu'] = kwargs.pop('nu', 0)
+        self.call_kwargs['ext'] = kwargs.pop('ext', None)
+
+        if fill_value is not None:
+            raise ValueError('SplineInterpolator does not support fill_value')
+
+        self.f = UnivariateSpline(xi, yi, k=order, **self.cons_kwargs)
+
+
+def _apply_over_vars_with_dim(func, self, dim=None, **kwargs):
+    '''wrapper for datasets'''
+    from .dataset import Dataset
+
+    ds = Dataset(coords=self.coords, attrs=self.attrs)
```

Comment:
If people are subclassing `Dataset` (we do internally), this will void their subclass. I think generally we should use `type(self)` or `self.__class__`